staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -178,7 +181,9 @@ func (cnc *CloudNodeController) Run(stopCh <-chan struct{}, controllerManagerMet
 		}
 	}, cnc.nodeStatusUpdateFrequency, stopCh)
 
-	go wait.Until(cnc.runWorker, time.Second, stopCh)
+	for i := int32(0); i < cnc.workerCount; i++ {
+		go wait.Until(cnc.runWorker, time.Second, stopCh)
```

Comment:
@cheftako have we verified everything the node controller does and spawns inside `syncNode` is threadsafe? this includes calls to the cloud provider interface, etc.

As a data point, the HPA controller also just added concurrency and there were several operations it was performing that were *not* threadsafe (https://github.com/kubernetes/kubernetes/pull/108501)

Reply:
this is the same pattern that was added recently to the node-lifecycle-controller

```diff
+++ b/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go
@@ -26,6 +26,7 @@ import (
        v1 "k8s.io/api/core/v1"
        apierrors "k8s.io/apimachinery/pkg/api/errors"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+       "k8s.io/apimachinery/pkg/labels"
        "k8s.io/apimachinery/pkg/types"
        utilruntime "k8s.io/apimachinery/pkg/util/runtime"
        "k8s.io/apimachinery/pkg/util/wait"
@@ -246,28 +247,36 @@ func (cnc *CloudNodeController) syncHandler(key string) error {
 
 // UpdateNodeStatus updates the node status, such as node addresses
 func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) error {
-       nodes, err := cnc.kubeClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{ResourceVersion: "0"})
+       start := nc.now()
+
+       nodes, err := cnc.nodeInformer.Lister().List(labels.Everything())
        if err != nil {
                klog.Errorf("Error monitoring node status: %v", err)
                return err
        }
 
-       for i := range nodes.Items {
-               instanceMetadata, err := cnc.getInstanceNodeAddresses(ctx, &nodes.Items[i])
-               if err != nil {
-                       klog.Errorf("Error getting instance metadata for node addresses: %v", err)
-                       continue
-               }
-               cnc.updateNodeAddress(ctx, &nodes.Items[i], instanceMetadata)
+       defer func() {
+               klog.V(2).Infof("Update %d nodes status took %v.", len(nodes),time.Since(start))
        }
 
-       for _, node := range nodes.Items {
-               err = cnc.reconcileNodeLabels(node.Name)
-               if err != nil {
-                       klog.Errorf("Error reconciling node labels for node %q, err: %v", node.Name, err)
+       updateNodeFunc := func(piece int) {
+               node := nodes[piece].DeepCopy()
+                       instanceMetadata, err := cnc.getInstanceNodeAddresses(ctx, node)
+                       if err != nil {
+                               klog.Errorf("Error getting instance metadata for node addresses: %v", err)
+                               continue
+                       }
+                       cnc.updateNodeAddress(ctx, &nodes.Items[i], instanceMetadata)
+               }
+
+                       err = cnc.reconcileNodeLabels(node.Name)
+                       if err != nil {
+                               klog.Errorf("Error reconciling node labels for node %q, err: %v", node.Name, err)
                }
        }
 
+       workqueue.ParallelizeUntil(ctx, cnc.workerCount, len(nodes), updateNodeFunc)
+
        return nil
 }
```

https://github.com/kubernetes/kubernetes/pull/114296/files#r1067096291

Reply:
I agree we also want to do that, but that's an orthogonal optimization.

This is a separate goroutine and it's not blocking the execution of anything else (it's just potentially slow). So I wouldn't necessary bundle this with this PR - let's do that in a separate PR.


Regarding this particulra change - I just reviewed that and:
- given that workqueue is thread-safe
- syncHandler is effectively calling syncNode
- reviewing carefully syncNode (and what it calls internally)
this is safe change.

Reply:
fair enough